crates/sui-core/src/authority_active/checkpoint_driver/mod.rs

```diff
@@ -370,6 +231,192 @@ where
+            );
+
+            // TODO: The sync process only works within an epoch.
+            sync_to_checkpoint(
+                active_authority.clone(),
+                state_checkpoints.clone(),
+                checkpoint.clone(),
+            )
+            .await
+            .map_err(|err| CheckpointStepError::SyncCheckpointFromQuorumFailed(Box::new(err)))?;
+
+            return Ok(CheckpointStepResult::NewCheckpointCertStored);
+        }
+    }
+    Err(CheckpointStepError::WaitForCheckpointCert)
+}
+
+async fn handle_checkpoint_step<A>(
+    result: Result<CheckpointStepResult, CheckpointStepError>,
+    active_authority: Arc<ActiveAuthority<A>>,
+    timing: &CheckpointProcessControl,
+    metrics: CheckpointMetrics,
+    last_cert_time: &mut Instant,
+) where
+    A: AuthorityAPI + Send + Sync + 'static + Clone + Reconfigurable,
+{
+    let state_checkpoints = &active_authority.state.checkpoints;
+    let next_cp_seq = state_checkpoints.lock().next_checkpoint();
+    match result {
+        Ok(result) => {
+            match result {
+                CheckpointStepResult::CheckpointSigned => {
+                    info!(
+                        ?next_cp_seq,
+                        "A new checkpoint is created and signed locally"
+                    );
+                    metrics.checkpoints_signed.inc();
+                }
+                CheckpointStepResult::NewCheckpointCertStored => {
+                    metrics
+                        .checkpoint_frequency
+                        .observe(last_cert_time.elapsed().as_secs_f64());
+                    metrics
+                        .checkpoint_sequence_number
+                        .set((next_cp_seq - 1) as i64);
+                    *last_cert_time = Instant::now();
+                    if active_authority.state.is_fullnode()
+                        && state_checkpoints.lock().is_ready_to_start_epoch_change()
+                    {
+                        while let Err(err) = active_authority.start_epoch_change().await {
+                            error!(?next_cp_seq, "Failed to start epoch change: {:?}", err);
+                            tokio::time::sleep(timing.epoch_change_retry_delay).await;
+                        }
+                        // No delay to minimize the reconfiguration latency.
+                        return;
+                    } else if active_authority.state.is_fullnode()
+                        && state_checkpoints.lock().is_ready_to_finish_epoch_change()
+                    {
+                        while let Err(err) = active_authority.finish_epoch_change().await {
+                            error!(?next_cp_seq, "Failed to finish epoch change: {:?}", err);
+                            tokio::time::sleep(timing.epoch_change_retry_delay).await;
+                        }
+                    }
+                    tokio::time::sleep(timing.long_pause_between_checkpoints).await;
```

Comment:
if i'm reading this right, fullnode will take a long pause after syncing every checkpoint. but we want fullnode to sync as fast as it can. But, when this is part of the checkpoint creation step (on validator), we do indeed want to pause after a new checkpoint is formed.

Reply:
Makes sense for new node startup syncing, but for incremental syncs of new checkpoints, won't removing this cause a ton of busy looping once there's a new checkpoint?